plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c

```diff
@@ -0,0 +1,332 @@
+		  OBS_VERSION);
+
+	vencoder = obs_output_get_video_encoder(stream->output);
+	settings = obs_encoder_get_settings(vencoder);
+	keyint_sec = obs_data_get_int(settings, "keyint_sec");
+	if (keyint_sec) {
+		dstr_catf(&stream->muxer_settings, " hls_time=%d", keyint_sec);
+		stream->keyint_sec = keyint_sec;
+	}
+
+	obs_data_release(settings);
+
+	start_pipe(stream, path.array);
+	dstr_free(&path);
+
+	if (!stream->pipe) {
+		obs_output_set_last_error(
+			stream->output, obs_module_text("HelperProcessFailed"));
+		warn("Failed to create process pipe");
+		return false;
+	}
+	stream->mux_thread_joinable = pthread_create(&stream->mux_thread, NULL,
+						     write_thread, stream) == 0;
+	if (!stream->mux_thread_joinable)
+		return false;
+
+	/* write headers and start capture */
+	os_atomic_set_bool(&stream->active, true);
+	os_atomic_set_bool(&stream->capturing, true);
+	os_atomic_set_bool(&stream->is_hls, true);
+	stream->total_bytes = 0;
+	stream->dropped_frames = 0;
+	stream->min_priority = 0;
+
+	obs_output_begin_data_capture(stream->output, 0);
+
+	dstr_copy(&stream->printable_path, path_str);
+	info("Writing to path '%s'...", stream->printable_path.array);
+	return true;
+}
+
+static bool write_packet_to_buf(struct ffmpeg_muxer *stream,
+				struct encoder_packet *packet)
+{
+	circlebuf_push_back(&stream->packets, packet,
+			    sizeof(struct encoder_packet));
+	return true;
+}
+
+static void drop_frames(struct ffmpeg_muxer *stream, int highest_priority)
+{
+	struct circlebuf new_buf = {0};
+	int num_frames_dropped = 0;
+
+	circlebuf_reserve(&new_buf, sizeof(struct encoder_packet) * 8);
+
+	while (stream->packets.size) {
+		struct encoder_packet packet;
+		circlebuf_pop_front(&stream->packets, &packet, sizeof(packet));
+
+		/* do not drop audio data or video keyframes */
+		if (packet.type == OBS_ENCODER_AUDIO ||
+		    packet.drop_priority >= highest_priority) {
+			circlebuf_push_back(&new_buf, &packet, sizeof(packet));
+		} else {
+			num_frames_dropped++;
+			obs_encoder_packet_release(&packet);
+		}
+	}
+
+	circlebuf_free(&stream->packets);
+	stream->packets = new_buf;
+
+	if (stream->min_priority < highest_priority)
+		stream->min_priority = highest_priority;
+	if (!num_frames_dropped)
+		return;
+
+	stream->dropped_frames += num_frames_dropped;
+}
+
+static bool find_first_video_packet(struct ffmpeg_muxer *stream,
+				    struct encoder_packet *first)
+{
+	size_t count = stream->packets.size / sizeof(*first);
+
+	for (size_t i = 0; i < count; i++) {
+		struct encoder_packet *cur =
+			circlebuf_data(&stream->packets, i * sizeof(*first));
+		if (cur->type == OBS_ENCODER_VIDEO && !cur->keyframe) {
+			*first = *cur;
+			return true;
+		}
+	}
+	return false;
+}
+
+void check_to_drop_frames(struct ffmpeg_muxer *stream, bool pframes)
+{
+	struct encoder_packet first;
+	int64_t buffer_duration_usec;
+	int priority = pframes ? OBS_NAL_PRIORITY_HIGHEST
+			       : OBS_NAL_PRIORITY_HIGH;
+	int64_t drop_threshold = 2 * stream->keyint_sec;
```

Comment:
I think we need to handle the case when keyint_sec may be zero. So we need a default threshold.